cli/azd/pkg/tools/azcli/keyvault.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/azure/azure-dev/cli/azd/pkg/identity"
+)
+
+type AzCliKeyVault struct {
+	Id         string `json:"id"`
+	Name       string `json:"name"`
+	Location   string `json:"location"`
+	Properties struct {
+		EnableSoftDelete      bool `json:"enableSoftDelete"`
+		EnablePurgeProtection bool `json:"enablePurgeProtection"`
+	} `json:"properties"`
+}
+
+type AzCliKeyVaultSecret struct {
+	Id    string `json:"id"`
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+func (cli *azCli) GetKeyVault(ctx context.Context, subscriptionId string, resourceGroupName string, vaultName string) (*AzCliKeyVault, error) {
+	client, err := cli.createKeyVaultClient(ctx, subscriptionId)
+	if err != nil {
+		return nil, err
+	}
+
+	vault, err := client.Get(ctx, resourceGroupName, vaultName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting keyvault: %w", err)
+	}
+
+	return &AzCliKeyVault{
+		Id:       *vault.ID,
+		Name:     *vault.Name,
+		Location: *vault.Location,
+		Properties: struct {
+			EnableSoftDelete      bool "json:\"enableSoftDelete\""
+			EnablePurgeProtection bool "json:\"enablePurgeProtection\""
+		}{
+			EnableSoftDelete:      convert.ToValueWithDefault(vault.Properties.EnableSoftDelete, false),
+			EnablePurgeProtection: convert.ToValueWithDefault(vault.Properties.EnablePurgeProtection, false),
+		},
+	}, nil
+}
+
+func (cli *azCli) GetKeyVaultSecret(ctx context.Context, vaultName string, secretName string) (*AzCliKeyVaultSecret, error) {
+	vaultUrl := vaultName
+	if !strings.Contains(strings.ToLower(vaultName), "https://") {
+		vaultUrl = fmt.Sprintf("https://%s.vault.azure.net", vaultName)
+	}
+
+	client, err := cli.createSecretsDataClient(ctx, vaultUrl)
+	if err != nil {
+		return nil, nil
+	}
+
+	response, err := client.GetSecret(ctx, secretName, "", nil)
+	if err != nil {
+		var httpErr *azcore.ResponseError
+
+		if errors.As(err, &httpErr) {
+			if httpErr.StatusCode == http.StatusNotFound {
+				return nil, ErrAzCliSecretNotFound
+			}
+		} else {
+			return nil, fmt.Errorf("getting key vault secret: %w", err)
+		}
+	}
+
+	return &AzCliKeyVaultSecret{
+		Id:    response.SecretBundle.ID.Version(),
+		Name:  response.SecretBundle.ID.Name(),
+		Value: *response.SecretBundle.Value,
+	}, nil
+}
+
+func (cli *azCli) PurgeKeyVault(ctx context.Context, subscriptionId string, vaultName string, location string) error {
+	client, err := cli.createKeyVaultClient(ctx, subscriptionId)
+	if err != nil {
+		return err
+	}
+
+	poller, err := client.BeginPurgeDeleted(ctx, vaultName, location, nil)
+	if err != nil {
+		return fmt.Errorf("starting purging key vault: %w", err)
+	}
+
+	_, err = poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("purging key vault: %w", err)
+	}
+
+	return nil
+}
+
+// Creates a KeyVault client for ARM control plane operations
+func (cli *azCli) createKeyVaultClient(ctx context.Context, subscriptionId string) (*armkeyvault.VaultsClient, error) {
+	cred, err := identity.GetCredentials(ctx)
```

Comment:
I keep wondering if we should store the credential object on the `AzCli` type itself, and then either have `NewAzCli` take credential directly (and have `GetAzCli` pull the credentail from the context and pass it in) or have `NewAzCli` take a context and pull the credential once there).

I'm wondering if that will make the work that @weikanglim is thinking about doing with Wire any easier.  My mental model is that the desing of `AzCli` over time will sort of mirror the "clients" that the Go SDK produces, and so the credential should be per client state instead of something we fetch per operation?

Reply:
The `identity.GetCredentials()` pulls the cred from context and only creates a new one if it does not already exist.  We're not setting the value in the context today so may make sense to do that now.

Reply:
I think an object (not `context)` that stores the credentials makes total sense.